dexpv2/segmentation.py

```diff
@@ -10,6 +11,135 @@
+              (array of float16) for all arrays, where the index
+              is the uint16 value and the value is the corresponding
+              original float16 value.
+
+    Raises:
+        TypeError: If any input array's dtype is not float16.
+        ValueError: If the list of arrays is empty, or if the total number
+                    of unique values across all arrays exceeds the capacity
+                    of uint16 (65536).
+    """
+    # Ensure xp is defined (this is more of a runtime check if not using static analysis)
+    if "xp" not in globals() and "xp" not in locals():
+        raise NameError(
+            "Array library 'xp' is not defined. Please import numpy as xp or cupy as xp."
+        )
+
+    if not float16_arrays:
+        raise ValueError("Input list of arrays cannot be empty.")
+
+    # Validate input types and collect original shapes and sizes
+    original_shapes = []
+    original_sizes = []
+    for i, arr in enumerate(float16_arrays):
+        # Assuming 'xp' is defined, xp.ndarray would be the type to check against
+        # For simplicity and following user's snippet, primarily checking dtype.
+        if not hasattr(arr, "dtype") or arr.dtype != xp.float16:
+            raise TypeError(
+                f"Array at index {i} must be an 'xp.ndarray' with dtype xp.float16. "
+                f"Got type {type(arr)} with dtype {getattr(arr, 'dtype', 'N/A')}."
+            )
+        original_shapes.append(arr.shape)
+        original_sizes.append(arr.size)
+
+    # Handle case where all arrays combined are empty
+    if sum(original_sizes) == 0:
+        # Create empty uint16 arrays with original shapes
+        empty_uint16_list = [
+            xp.array([], dtype=xp.uint16).reshape(shape) for shape in original_shapes
+        ]
+        return empty_uint16_list, xp.array([], dtype=xp.float16)
+
+    # Concatenate all arrays into a single flat array for global unique value finding.
+    # We need to ensure that we only concatenate non-empty arrays if ravel() on empty
+    # arrays with certain shapes causes issues, or handle shapes carefully.
+    # xp.concatenate([arr.ravel() for arr in float16_arrays]) should generally work.
+    # Ravel ensures that even multi-dimensional arrays become 1D before concatenation.
+    try:
+        combined_float16_array = xp.concatenate([arr.ravel() for arr in float16_arrays])
+    except Exception as e:
+        raise ValueError(
+            f"Error during concatenation of arrays: {e}. Ensure 'xp' is correctly defined (NumPy/CuPy)."
+        )
+
+    # Find unique values and their inverse indices from the combined array.
+    # unique_values will be sorted, which is crucial for order preservation.
+    # inverse_indices will correspond to the flattened combined_float16_array.
+    unique_values: "xp.ndarray"  # type: ignore
+    inverse_indices: "xp.ndarray"  # type: ignore
+    unique_values, inverse_indices = xp.unique(
```

Comment:
```suggestion
    unique_values, inverse_indices = np.unique(
```
we could start favoring using `np` dispatch mechanism, I'm not sure if I'm doing this everywhere